### PR TITLE
kn source apiserver create/delete  (as typed-builtin source)

### LINKED ISSUE #334: Add "kn source" create & update typed builtin
## In what area(s)?

Eventing

## Describe the feature:

As [MVP](https://en.wikipedia.org/wiki/Minimum_viable_product) we suggest implementing all use cases for kn source command as this is the smallest required action for connecting a given Knative service to an event producer. It does not require any other parts of Knative eventing but only Knative sources installed.

The MVP should contain built-in support for 

* CronJobSource
* ApiServerSource
* ContainerSource

For the suggested command line syntax, please refer to [Sources](https://docs.google.com/document/d/1B8VK-5nuxt3D-i1Up3onmUbzcztQXE0GRgf1lfx49P8/edit#heading=h.8yhces3mysbr) in the proposed spec.

### CronJobSource

```
kn sources cron create mycron --schedule "* * * * 5/*" --data "ping" --sink myservice

kn sources cron update mycron --schedule "* * * * 5/*"
kn sources cron update mycron --remove-sink

kn sources cron delete mycron

kn sources cron list mycron
```

### ApiServerSource

```
kn sources api-server create myapiserver --resource pod --resource deployment --mode ref

....
```

### ContainerSource

```
kn sources container create myimporter --image gcr/test-importer --arg firstArg --arg secondArg --env var1=value1 --env var2=value2

....
```

* [Overview slides](https://docs.google.com/presentation/d/1rSoclSuzV_QAyyvD6jIpbOfN-OrFvUyC7yWI54SfnE0/edit?usp=sharing)
* ~~[Knative eventing support in Kn - V3](https://docs.google.com/document/d/19MO8JBPU5JrBthwI5jYs_V0zzgMtTmiOHou6w_T04XI/edit?usp=sharing)~~ replaced ny [Knative eventing support in Kn - Source & Trigger - V4](https://docs.google.com/document/d/1B8VK-5nuxt3D-i1Up3onmUbzcztQXE0GRgf1lfx49P8/edit?usp=sharing)

### PR DESCRIPTION
**Pull request by daisy-ycguo**
This is a second builtin-typed source as described in [specification](https://docs.google.com/document/d/1B8VK-5nuxt3D-i1Up3onmUbzcztQXE0GRgf1lfx49P8/edit#bookmark=id.d382ob8kwx0f), in addition to the cron-job source ib #357  

Partially address #499

## Proposed Changes

* Add command to create ApiServerSource
* Add command delete ApiServerSource

### CODE REVIEW

File: pkg/kn/commands/source/apiserver/apiserver_flags.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
+// Copyright © 2019 The Knative Authors
```
Comment (navidshaikh) [inline pkg/kn/commands/source/apiserver/apiserver_flags.go]: can we move the flags to `pkg/kn/commands/flags/` ?

File: pkg/kn/commands/source/sink/sink_flags.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
+// Copyright © 2019 The Knative Authors
```
Comment (navidshaikh) [inline pkg/kn/commands/source/sink/sink_flags.go]: can we move the flags to `pkg/kn/commands/flags/` ?

File: pkg/kn/commands/source/source_flags.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// Copyright © 2019 The Knative Authors
```
Comment (navidshaikh) [inline pkg/kn/commands/source/source_flags.go]: can we move the flags to `pkg/kn/commands/flags/` ?

**Comment by toVersus** [comment]
@rhuss As described below, you (co-author of this PR) need to comment `@googlebot I consent`.

> We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that by leaving a comment that contains only `@googlebot I consent`. in this pull request.

File: docs/cmd/kn_source_api-server_create.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+```
+
+  # Create a Kubernetes Api Server event source 'k8sevents' which consumes Kubernetes events and send them to service 'mysvc' as a cloudevent
+  kn source api-server create k8sevents --resources Event --accountname myaccountname --sink svc:mysvc
+```
+
+### Options
+
+```
+      --accountName string   Name of the ServiceAccount to use to run this source
```
Comment (maximilien) [inline docs/cmd/kn_source_api-server_create.md]: Inconsistent use of camel case between `accountName` and `apiversion`.

Unless there is a strong reason, I'd suggest to be consistent and use camel case in all cases.

File: pkg/kn/commands/source/apiserver/apiserver_create.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	"knative.dev/client/pkg/kn/commands"
+	"knative.dev/client/pkg/kn/commands/source/sink"
+)
+
+func NewApiServerCreateCommand(p *commands.KnParams) *cobra.Command {
+	var apiServerUpdateFlags ApiServerSourceUpdateFlags
+	var sinkFlags sink.SinkFlags
+
+	cmd := &cobra.Command{
+		Use:   "create NAME --resources RESOURCE --accountname ACCOUNTNAME --sink SINK --mode MODE",
```
Comment (maximilien) [inline pkg/kn/commands/source/apiserver/apiserver_create.go]: `--accountName` ?

File: pkg/kn/commands/source/apiserver/apiserver_create.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+			servingClient, err := p.NewClient(namespace)
+			if err != nil {
+				return err
+			}
+
+			config, err := p.GetClientConfig()
+			if err != nil {
+				return err
+			}
+			aipServerSourceClient, err := NewEventingClient(config, namespace)
```
Comment (maximilien) [inline pkg/kn/commands/source/apiserver/apiserver_create.go]: Where is the check for `err` after line 57? What if the `NewEventingClient` fails? If it can never fail then don't have it return an error.

File: pkg/kn/commands/source/apiserver/apiserver_create.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+			objectRef, err := sinkFlags.ResolveSink(servingClient)
+			if err != nil {
+				return err
+			}
+			return aipServerSourceClient.CreateApiServerSource(name, apiServerUpdateFlags.getApiServerResourceArray(), apiServerUpdateFlags.accountName, apiServerUpdateFlags.mode, objectRef)
```
Comment (maximilien) [inline pkg/kn/commands/source/apiserver/apiserver_create.go]: This reads odd and could be improve a bit by unpacking the return. For instance:

```go
resourceArray := apiServerUpdateFlags.getApiServerResourceArray()
serverSource := aipServerSourceClient.CreateApiServerSource(name, resourceArray, apiServerUpdateFlags.accountName, apiServerUpdateFlags.mode, objectRef)
return serverSource
```

Also can any of these `CreateApiXxx(...)` call produce errors?

File: pkg/kn/commands/source/apiserver/apiserver_flags.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+func getValidResource(resource string) (string, string, bool) {
+	var version, kind string
+	var isController bool
+	parts := strings.Split(resource, "/")
+	if len(parts) == 3 {
+		version = parts[0]
+		kind = parts[1]
+		isController, _ = strconv.ParseBool(parts[2])
```
Comment (maximilien) [inline pkg/kn/commands/source/apiserver/apiserver_flags.go]: Why are we ignoring errors? [`strconv.ParseBool`](https://golang.org/pkg/strconv/#ParseBool) returns an error if the string passed is not a bool. Should deal with error.

File: pkg/kn/commands/source/source.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+	"knative.dev/client/pkg/kn/commands"
+	"knative.dev/client/pkg/kn/commands/source/apiserver"
+)
+
+func NewImporterCommand(p *commands.KnParams) *cobra.Command {
+	sourceCmd := &cobra.Command{
+		Use:   "source",
+		Short: "Event Source command group",
+	}
+	//sourceCmd.AddCommand(cron.NewCronCommand(p))
```
Comment (maximilien) [inline pkg/kn/commands/source/source.go]: Don't leave out commented code. Just remove it if not needed.

**Comment by daisy-ycguo** [comment]
Thanks for all the comments. Working on them now.

**Comment by daisy-ycguo** [comment]
@googlebot I consent

**Comment by rhuss** [comment]
@googlebot I consent

File: pkg/kn/commands/source/apiserver/apiserver_create.go (hunk 5 above)
Comment (rhuss) [reply pkg/kn/commands/source/apiserver/apiserver_create.go]: whyt not just `--account` ?

Comment (rhuss) [reply pkg/kn/commands/source/apiserver/apiserver_create.go]: `--resource` instead of `--resources` ?

**Comment by daisy-ycguo** [comment]
Thanks for comments. I will adopt them. Just add a few explanations. 

- `--mode`, mode is described [here](https://github.com/knative/eventing/blob/master/pkg/apis/sources/v1alpha1/apiserver_types.go#L90) in Eventing [apiserver_types.go](https://github.com/knative/eventing/blob/master/pkg/apis/sources/v1alpha1/apiserver_types.go). I don't quite get the point, just copying the explanation here. Maybe you can help to make a better description of `mode`.
- `--resources`, here is the [definition](https://github.com/knative/eventing/blob/master/pkg/apis/sources/v1alpha1/apiserver_types.go#L106) of resources in Eventing apiserver_types.go. Again, `Controller` is a boolean and the description is not clear here. I changed it to `isController` just hoping to give a clear hint that the type is boolean. `--resources` is used to describe the K8s resources that you want to monitor to. Of course users can monitor multiple resources all together. I plan to use something like `Event\v1\true,Pod\v1\false` to describe resources.

ApiServerResource is complicated than cronjob and ContainerSource. Welcome to any suggestions to get a better UX. Here is a [sample](https://knative.dev/docs/eventing/samples/kubernetes-event-source/) of ApiServerSource, which can help to get a better understanding.

File: pkg/kn/commands/source/apiserver/apiserver_client.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
+// Copyright © 2019 The Knative Authors
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/apiserver_client.go]: done

File: pkg/kn/commands/source/apiserver/apiserver.go [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"github.com/spf13/cobra"
+
+	"knative.dev/client/pkg/kn/commands"
+)
+
+func NewApiServerCommand(p *commands.KnParams) *cobra.Command {
+	apiServerImporterCmd := &cobra.Command{
+		Use:   "api-server",
+		Short: "Kubernetes Api Server Event Source command group",
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/apiserver.go]: done

File: hack/build.sh [hunk 12]
```diff
@@ -263,13 +263,9 @@ has_flag() {
...
-cross_build() {
-  local basedir=$(basedir)
-  local ld_flags="$(build_flags $basedir)"
-  local pkg="github.com/knative/client/pkg/kn/commands"
-  local failed=0
-
-  echo "⚔️ ${S}Compile"
+build_cross() {
+  local ld_flags="$(build_flags $(dirname $0)/..)"
+  local pkg="knative.dev/client/pkg/kn/commands"
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply hack/build.sh]: done

File: pkg/kn/commands/source/apiserver/apiserver_flags.go (hunk 1 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/apiserver_flags.go]: These are flags for `kn source create apiserver` only.
I think it's better to keep it in `commands/source/apiserver` folder.

File: pkg/kn/commands/source/sink/sink_flags.go (hunk 2 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/sink/sink_flags.go]: Maybe sink could be moved to `pkg/kn/commands/flags/` because it will be used in several Eventing command. I will leave it to @rhuss to decide.

File: pkg/kn/commands/source/source_flags.go (hunk 3 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/source_flags.go]: I will suggest we keep source flags only in `kn/commands/source` folder, because it is `source` specific.

File: docs/cmd/kn_source_api-server_create.md (hunk 4 above)
Comment (daisy-ycguo) [reply docs/cmd/kn_source_api-server_create.md]: done

File: pkg/kn/commands/source/apiserver/apiserver_create.go (hunk 5 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/apiserver_create.go]: I plan to use `--resources` to describe a list of resources because ApiServerSource supports to define the monitoring of multiple resources.

File: pkg/kn/commands/source/apiserver/apiserver_flags.go (hunk 1 above)
Comment (navidshaikh) [reply pkg/kn/commands/source/apiserver/apiserver_flags.go]: It's more about having the kn flags together in flags package, right now not all commands flags are following it (only traffic flags). We've an open issue for this https://github.com/knative/client/issues/337.

File: pkg/kn/commands/source/apiserver/apiserver_create.go (hunk 6 above)
Comment (maximilien) [reply pkg/kn/commands/source/apiserver/apiserver_create.go]: Please address ^^^

File: pkg/kn/commands/source/apiserver/apiserver_flags.go (hunk 8 above)
Comment (maximilien) [reply pkg/kn/commands/source/apiserver/apiserver_flags.go]: Please address ^^^

File: pkg/kn/commands/source/source.go (hunk 9 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/source.go]: done

File: pkg/kn/commands/source/apiserver/apiserver_create.go (hunk 6 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/apiserver_create.go]: done

File: pkg/kn/commands/source/apiserver/apiserver_flags.go (hunk 8 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/apiserver_flags.go]: done

File: pkg/kn/commands/source/apiserver/apiserver_create.go (hunk 7 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/apiserver_create.go]: CreateApiServerSource could produce error. In the current code base, it has been changed to `return aipServerSourceClient.CreateApiServerSource(...)`. So the error will be returned if it is not nil.

File: pkg/kn/commands/source/source_flags.go (hunk 3 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/source_flags.go]: I moved all sources flags to pkg/kn/commands/flags/

File: pkg/kn/commands/source/sink/sink_flags.go (hunk 2 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/sink/sink_flags.go]: I have moved flags to that folder.

File: pkg/kn/commands/source/apiserver/apiserver_flags.go (hunk 1 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/apiserver_flags.go]: I have moved flags to that folder.

**Comment by daisy-ycguo** [comment]
I'm working on the testings and will add more tests in next commit.

File: pkg/kn/commands/types_test.go [hunk 13]
```diff
@@ -141,3 +141,53 @@ func TestGetClientConfig(t *testing.T) {
...
+			"",
+			true,
+		},
+	} {
+		p := &KnParams{
+			ClientConfig: tc.clientConfig,
+			LogHTTP:      tc.logHttp,
+		}
+
+		sourcesClient, err := p.newSourcesClient(namespace)
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply pkg/kn/commands/types_test.go]: follow up suggestion based on above comment

[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply pkg/kn/commands/types_test.go]: +1

File: pkg/kn/commands/types.go [hunk 14]
```diff
@@ -40,10 +42,11 @@ type Config struct {
...
 type KnParams struct {
-	Output       io.Writer
-	KubeCfgPath  string
-	ClientConfig clientcmd.ClientConfig
-	NewClient    func(namespace string) (serving_kn_v1alpha1.KnServingClient, error)
+	Output           io.Writer
+	KubeCfgPath      string
+	ClientConfig     clientcmd.ClientConfig
+	NewClient        func(namespace string) (serving_kn_v1alpha1.KnServingClient, error)
+	NewSourcesClient func(namespace string) (sources_kn_v1alpha1.KnSourcesClient, error)
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply pkg/kn/commands/types.go]: agree. Thanks.

**Comment by daisy-ycguo** [comment]
I split the common codes into a separate PR #514, for easy collaboration among all sources client features.

**Comment by navidshaikh** [comment]
e2e flake should be fixed by https://github.com/knative/client/pull/523

**Comment by rhuss** [comment]
This PR is on my bucket list for today. It's a big list :)  but for sure I will review the PR until tomorrow.

**Comment by navidshaikh** [comment]
Ran the PR locally, few things to note:
```
 ./kn source apiserver create k8sevents --resources Event:v1:true --service-account myaccountname --sink svc:mysvc

services.serving.knative.dev "mysvc" not found
```
 - (optional) We can improve this message, but it's not consistent even between list or describe

```
➜  client git:(apis) kubectl get apiserversource k8sevents 
NAME        READY   AGE
k8sevents   False   8m21s

➜  client git:(apis) ./kn source apiserver create k8sevents --resources Event:v1:true  --sink svc:hello

ApiServerSource 'k8sevents' successfully created in namespace 'default'.
apiserversources.sources.eventing.knative.dev "k8sevents" already exists
```
 - Creation error message is printed before existence check could return.

```
./kn source apiserver create --help
[..]
      --resources string         comma seperate Kind:APIVersion:isController list, e.g. Event:v1:true (default "r")
```
 - We should rephrase this to "Comma separated Kind:APIVersion:isController list, e.g. Event:v1:true (default "r")"  ?

File: pkg/kn/commands/source/apiserver/create.go [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+			if err != nil {
+				return err
+			}
+
+			// construct ApiServerSource
+			apisrvsrc := constructApiServerSource(name, namespace, apiServerUpdateFlags)
+			apisrvsrc.Spec.Sink = objectRef
+
+			// create
+			_, err = sourcesClient.CreateApiServerSource(apisrvsrc)
```
Comment (navidshaikh) [inline pkg/kn/commands/source/apiserver/create.go]: error check missing

File: pkg/kn/commands/source/apiserver/create_flag.go [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	return version, kind, isController
+}
+
+//AddApiServerFlags is to set parameters
+func (f *ApiServerSourceUpdateFlags) Add(cmd *cobra.Command) {
+	cmd.Flags().StringVar(&f.ServiceAccountName, "service-account", "", "Name of the ServiceAccount to use to run this source")
+	cmd.Flags().StringVar(&f.Mode, "mode", "Ref", "'Ref' or 'Resource', the receive adapter controller runs under.")
+	cmd.Flags().StringVar(&f.ApiVersion, "apiVersion", "1.0", "Kubernetes API version of the resource to watch")
+	cmd.Flags().StringVar(&f.Resources, "resources", "r", "comma seperate Kind:APIVersion:isController list, e.g. Event:v1:true")
```
Comment (navidshaikh) [inline pkg/kn/commands/source/apiserver/create_flag.go]: should add an example like documented in code `Event:v1:true,Pod:v2:false`

File: pkg/kn/commands/source/apiserver/create_flag.go [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/spf13/cobra"
+	sources_v1alpha1 "knative.dev/eventing/pkg/apis/sources/v1alpha1"
+)
+
+const ApiVersionSplitStr = ":"
```
Comment (navidshaikh) [inline pkg/kn/commands/source/apiserver/create_flag.go]: `APIVersionSplitChar`

File: pkg/kn/commands/source/apiserver/create_flag.go [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+const ApiServerSourceArraySplitStr = ","
```
Comment (navidshaikh) [inline pkg/kn/commands/source/apiserver/create_flag.go]: *APIServerSourceArraySplitChar

File: pkg/kn/commands/source/apiserver/create_flag.go [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	return resourceList
+}
+
+func getValidResource(resource string) (string, string, bool) {
+	var version, kind string
+	var isController bool
+	var err error
+	parts := strings.Split(resource, ApiVersionSplitStr)
+	if len(parts) == 3 {
```
Comment (navidshaikh) [inline pkg/kn/commands/source/apiserver/create_flag.go]: should we check for other cases of len as 2  or 1 ?

File: pkg/kn/commands/source/apiserver/create.go (hunk 15 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/create.go]: fixed

File: pkg/kn/commands/source/apiserver/create_flag.go (hunk 16 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/create_flag.go]: done

File: pkg/kn/commands/source/apiserver/create_flag.go (hunk 18 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/create_flag.go]: done

File: pkg/kn/commands/source/apiserver/create_flag.go (hunk 17 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/create_flag.go]: done

File: pkg/kn/commands/source/apiserver/create_flag.go (hunk 19 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/create_flag.go]: use default value to set Version and isController, in order to support other cases as 2 or 1.

**Comment by daisy-ycguo** [comment]
Thanks for review. Here are the updates in the latest version.

> Ran the PR locally, few things to note:
> 
> ```
>  ./kn source apiserver create k8sevents --resources Event:v1:true --service-account myaccountname --sink svc:mysvc
> 
> services.serving.knative.dev "mysvc" not found
> ```
> 
> * (optional) We can improve this message (though the error message is not consistent even between list or describe)
Updated to:
```
$ ./kn source apiserver create k8sevents --resources Event:v1:true --service-account service-sa --sink svc:mysvc
cannot create ApiServerSource 'k8sevents' in namespace 'default' because services.serving.knative.dev "mysvc" not found
```
> 
> ```
> ➜  client git:(apis) kubectl get apiserversource k8sevents 
> NAME        READY   AGE
> k8sevents   False   8m21s
> 
> ➜  client git:(apis) ./kn source apiserver create k8sevents --resources Event:v1:true  --sink svc:hello
> 
> ApiServerSource 'k8sevents' successfully created in namespace 'default'.
> apiserversources.sources.eventing.knative.dev "k8sevents" already exists
> ```
> * Creation error message is printed before existence check could return.

Updated:
```
$ k get apiserversource
NAME             AGE
k8sevents        5m15s
$ ./kn source apiserver create k8sevents --resources Event:v1:true --service-account service-sa --sink svc:fib-knative
cannot create ApiServerSource 'k8sevents' in namespace 'default' because apiserversources.sources.eventing.knative.dev "k8sevents" already exists
```
> 
> ```
> ./kn source apiserver create --help
> [..]
>       --resources string         comma seperate Kind:APIVersion:isController list, e.g. Event:v1:true (default "r")
> ```
> 
> * We should rephrase this to "Comma separated Kind:APIVersion:isController list, e.g. Event:v1:true (default "r")"  ?
Updated:
```
$ ./kn source apiserver create --help
Create a Kubernetes API Server event source

Usage:
  kn source apiserver create NAME --resources RESOURCE --service-account ACCOUNTNAME --sink SINK --mode MODE [flags]

......
Flags:
......
      --resources string         comma seperate
                                 Kind:APIVersion:isController list, e.g. Event:v1:true,Pod:v2:false
```

File: pkg/eventing/sources/v1alpha1/client.go [hunk 20]
```diff
@@ -23,6 +27,10 @@ import (
 type KnSourcesClient interface {
 	// Namespace in which this client is operating for
 	Namespace() string
+
```
Comment (rhuss) [inline pkg/eventing/sources/v1alpha1/client.go]: For building sources, I wonder whether we should add an intermediate layer (like having an `KnApiServerSourceClient` which can be accessed from this interface). This would easily help in separation of concerns (as it this specialized client could be stored and picked up separately by the using code). This of course depends also on the amount of sources we expect, but even for 3 sources, I think this would be helpful (this is also how the generated service client does it).

I'm totally fine to keep it like this for the current PR but would probably refactor in a subsequent PR, when we e.. introduce the CronJobSource.

File: pkg/kn/commands/flags/sink.go [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+func (i *SinkFlags) Add(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&i.sink, "sink", "s", "", "Addressable sink for events")
+}
+
+func (i *SinkFlags) ResolveSink(client v1alpha1.KnServingClient) (*apisv1alpha1.Destination, error) {
+	if i.sink == "" {
+		return nil, nil
+	}
+
+	if strings.HasPrefix(i.sink, "svc:") {
```
Comment (rhuss) [inline pkg/kn/commands/flags/sink.go]: Good start, we will need to add more sink types (as specified) later. I'm currently working on it as part of the CronJob PR, so totally fine with it here like it is.

File: docs/cmd/kn_source_apiserver_create.md [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+```
+
+### Options
+
+```
+  -h, --help                     help for create
+      --mode string              'Ref' or 'Resource', the receive adapter controller runs under. (default "Ref")
+  -n, --namespace string         Specify the namespace to operate in.
+      --resources string         comma seperate Kind:APIVersion:isController list, e.g. Event:v1:true,Pod:v2:false
+      --service-account string   Name of the ServiceAccount to use to run this source
```
Comment (rhuss) [inline docs/cmd/kn_source_apiserver_create.md]:
```suggestion
      --service-account string   Name of the service account to use for running this source
```

(for aliging with `kn service` commands

File: pkg/kn/commands/source/apiserver/create.go [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+			if err != nil {
+				return fmt.Errorf(
+					"cannot create ApiServerSource '%s' in namespace '%s' "+
+						"because %s", name, namespace, err)
```
Comment (rhuss) [inline pkg/kn/commands/source/apiserver/create.go]: You'd better use `%v` instead of `%s` when formatting an error.

File: pkg/kn/commands/source/apiserver/create_flag.go [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+			Kind:       kind,
+			Controller: controller,
+		}
+		resourceList = append(resourceList, resourceRef)
+	}
+	return resourceList
+}
+
+func getValidResource(resource string) (string, string, bool) {
+	var version = DefaultApiVersion // v1 as default
```
Comment (rhuss) [inline pkg/kn/commands/source/apiserver/create_flag.go]: would call it `apiGroupVersion` as this is supposed be the api-group + version combined (like in in `apps/v1` or `serving.knative.dev/v1alpha`)

File: pkg/kn/commands/source/apiserver/create_flag.go [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+)
+
+// ApiServerSourceUpdateFlags are flags for create and update a ApiServerSource
+type ApiServerSourceUpdateFlags struct {
+	ServiceAccountName string
+	Mode               string
+	Resources          string
+}
+
+// GetApiServerResourceArray is to return an array of ApiServerResource from a string. A sample is Event:v1:true,Pod:v2:false
```
Comment (rhuss) [inline pkg/kn/commands/source/apiserver/create_flag.go]: Please don't use `Pod:v2` as an example as this does not exist. Instead use also an example with an apiGroup (like in `Deployment:apps/v1`)

File: pkg/kn/commands/source/apiserver/create_flag.go (hunk 24 above)
Comment (rhuss) [reply pkg/kn/commands/source/apiserver/create_flag.go]: Also, IMO the version should be mandatory. Having `v1` as a default doesn't not make sense except for the most basice objects which does not carry an API group.

File: pkg/kn/commands/source/apiserver/create_flag.go [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+func (f *ApiServerSourceUpdateFlags) GetApiServerResourceArray() []sources_v1alpha1.ApiServerResource {
+	var resourceList []sources_v1alpha1.ApiServerResource
+	for _, r := range strings.Split(f.Resources, ApiServerSourceArraySplitChar) {
```
Comment (rhuss) [inline pkg/kn/commands/source/apiserver/create_flag.go]: I'm not sure if already had that discussion, but I would much prefer to allow multiple options for `--resource` instead of cramming everything into a single one, comma separated. It should be modelled much like `--traffic` on `kn service update`, to be consistent and and e.g. also for clarity.

File: docs/cmd/kn_source_apiserver_create.md [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+  kn source apiserver create k8sevents --resources Event:v1:true --service-account myaccountname --sink svc:mysvc
+```
+
+### Options
+
+```
+  -h, --help                     help for create
+      --mode string              'Ref' or 'Resource', the receive adapter controller runs under. (default "Ref")
+  -n, --namespace string         Specify the namespace to operate in.
+      --resources string         comma seperate Kind:APIVersion:isController list, e.g. Event:v1:true,Pod:v2:false
```
Comment (rhuss) [inline docs/cmd/kn_source_apiserver_create.md]: Please use a realistic example and maybe avoid the comma separated approach. See my comments below please.

File: docs/cmd/kn_source_apiserver_create.md [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+
+  # Create a Kubernetes API Server event source 'k8sevents' which consumes Kubernetes events and send them to service 'mysvc' as a cloudevent
+  kn source apiserver create k8sevents --resources Event:v1:true --service-account myaccountname --sink svc:mysvc
+```
+
+### Options
+
+```
+  -h, --help                     help for create
+      --mode string              'Ref' or 'Resource', the receive adapter controller runs under. (default "Ref")
```
Comment (rhuss) [inline docs/cmd/kn_source_apiserver_create.md]: Could you briefly explain here the difference between `Ref` and `Resource` ?

File: docs/cmd/kn_source_apiserver_create.md (hunk 27 above)
Comment (rhuss) [reply docs/cmd/kn_source_apiserver_create.md]: What does the boolean `controller` flag mean? Can we add this here into the documentation? Also please add which parts are mandatory and which are optional.

If this is too long then add the explanation outside of the options to the top-level command doc.

File: docs/cmd/kn_source_apiserver_create.md [hunk 29]
```diff
@@ -0,0 +1,43 @@
+## kn source apiserver create
+
+Create a Kubernetes API Server event source
+
+### Synopsis
+
+Create a Kubernetes API Server event source
+
```
Comment (rhuss) [inline docs/cmd/kn_source_apiserver_create.md]: Here you could explain a bit the different modes and what a controller is, which kinds of events are delivered (== all), etc.

File: pkg/kn/commands/source/apiserver/delete.go [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+			sourcesClient, err := p.NewSourcesClient(namespace)
+			if err != nil {
+				return err
+			}
+
+			err = sourcesClient.DeleteApiServerSource(name)
+			if err != nil {
+				return err
+			}
+			fmt.Fprintf(cmd.OutOrStdout(), "ApiServerSource '%s' successfully deleted in namespace '%s'.\n", args[0], namespace)
```
Comment (rhuss) [inline pkg/kn/commands/source/apiserver/delete.go]:
```suggestion
			fmt.Fprintf(cmd.OutOrStdout(), "ApiServerSource '%s' deleted in namespace '%s'.\n", args[0], namespace)
```

File: pkg/eventing/sources/v1alpha1/client.go [hunk 31]
```diff
@@ -23,6 +27,10 @@ import (
...
+	CreateApiServerSource(apisvrsrc *v1alpha1.ApiServerSource) (*v1alpha1.ApiServerSource, error)
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply pkg/eventing/sources/v1alpha1/client.go]: addressed.

File: pkg/eventing/sources/v1alpha1/client_test.go [hunk 32]
```diff
@@ -0,0 +1,108 @@
...
+	return
+}
+
+func TestDeleteApiServerSource(t *testing.T) {
+	var srcName = "new-src"
+	sourcesServer, client := setup()
+
+	apisourceNew := newApiServerSource(srcName)
+
+	sourcesServer.AddReactor("create", "apiserversources",
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply pkg/eventing/sources/v1alpha1/client_test.go]: Thanks. I will create an issue to track.

File: pkg/eventing/sources/v1alpha1/client.go (hunk 20 above)
Comment (daisy-ycguo) [reply pkg/eventing/sources/v1alpha1/client.go]: Thanks.

File: pkg/kn/commands/flags/sink.go (hunk 21 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/flags/sink.go]: Agree.

File: docs/cmd/kn_source_apiserver_create.md (hunk 22 above)
Comment (daisy-ycguo) [reply docs/cmd/kn_source_apiserver_create.md]: fixed in create_flag.go

File: pkg/kn/commands/source/apiserver/create_flag.go (hunk 26 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/create_flag.go]: `--traffic` is `StringSliceVar`, for example `v1=50,v2=50`; I will follow `--traffic` to handle `--resource`.

File: pkg/kn/commands/source/apiserver/delete.go (hunk 30 above)
Comment (daisy-ycguo) [reply pkg/kn/commands/source/apiserver/delete.go]: fixed

File: pkg/eventing/sources/v1alpha1/client_test.go (hunk 32 above)
[the remark this answers is not in the document]
Comment (rhuss) [reply pkg/eventing/sources/v1alpha1/client_test.go]: On a second thought having a dedicated mock implementation introduced only for the source looks like overkill. In fact for the CronJob source I even went so far to make it package private to reduce coupling (i.e. do not introduce the client into the KnCommand). Still fiddling for how to get the fake client injected into the test, but I think the more self-contained we get the source implementations (without having top-level hooks like a global KnSourcesClient), the better, e.g. for latter refactoring (e.g. when the ApiServer or CronJob sources will move out of eventing which is not unlikely).

**Comment by daisy-ycguo** [comment]
@googlebot I consent.

**Comment by rhuss** [comment]
> @googlebot I consent.

Hmm, seems not to be picked up. wonder whether this is a bot issue or more general.

File: docs/cmd/kn_source_apiserver_create.md (hunk 29 above)
Comment (daisy-ycguo) [reply docs/cmd/kn_source_apiserver_create.md]: done

File: docs/cmd/kn_source_apiserver_create.md (hunk 28 above)
Comment (daisy-ycguo) [reply docs/cmd/kn_source_apiserver_create.md]: done.

File: docs/cmd/kn_source_apiserver_create.md (hunk 27 above)
Comment (daisy-ycguo) [reply docs/cmd/kn_source_apiserver_create.md]: done

**Comment by daisy-ycguo** [comment]
@rhuss The major changes in the last version:
- follow `--traffic` to manage flag `--resource`
- add enough documentation to explain the flags
- minor changes.

Please review again. Thank you.

**Comment by rhuss** [comment]
@daisy-ycguo thanks ! There are some minor things to fix, but let's save a round-trip an merge now. If you don't mind I would update the minor fixes (like doc wording etc) in an extra PR.

/lgtm
/approve
